File: ruby/jelly/jelly.rb
module Jelly
    class JellyShape
        @@shapes = Hash.new()

        def self.register_shape(positions)
            positions.sort!
            positions.freeze()
            unless @@shapes.has_key?(positions)
                shape = JellyShape.new(positions)
                shape.freeze()
                @@shapes[positions] = shape
            end
            return @@shapes[positions]
        end

        attr_accessor :positions
        attr_accessor :w, :h
        attr_accessor :lines, :adjacent_lines  # bitboard

        def initialize(positions = [])
            @positions = positions
            @w = positions.map {|pos| pos[0]}.max + 1
            @h = positions.map {|pos| pos[1]}.max + 1

            lines = Array.new(@h) { 0 }
            @positions.each do |pos|
                lines[pos[1]] |= 1 << pos[0]
            end
            @lines = lines

            adjacent_lines = Array.new(@h + 2) { 0 }
            lines.each_with_index do |line, i|
                adjacent_lines[i    ] |= line << 1
                adjacent_lines[i + 1] |= line | (line << 2)
                adjacent_lines[i + 2] |= line << 1
            end
            @adjacent_lines = adjacent_lines
        end

        def occupy_position?(x, y)
            return @positions.include?([x, y])
        end

        def concatenated(shape, dx, dy)
            positions = @positions.map(&:dup)
            if dx < 0
                positions.each do |pos|
                    pos[0] -= dx
                end
                dx = 0
            end
            if dy < 0
                positions.each do |pos|
                    pos[1] -= dy
                end
                dy = 0
            end

            shape.positions.each do |pos|
                positions << [pos[0] + dx, pos[1] + dy]
            end
            return JellyShape.register_shape(positions)
        end

        def inspect
            return "#<JellyShape: w=#{w},h=#{h} positions=#{@positions}>"
        end

        def hash()
            return @positions.hash
        end

        def eql?(other)
            return @positions == other.positions
        end
    end

    class Jelly
        attr_accessor :x, :y, :color, :shape
        attr_accessor :locked
        attr_accessor :black_char
        attr_accessor :link_prev, :link_next
        attr_accessor :hiddens

        def initialize(x, y, color, shape, locked = false)
            @x = x
            @y = y
            @color = color
            @shape = shape
            @locked = locked
            @link_prev = @link_next = nil
            @hiddens = nil
        end

        def occupy_position?(x, y)
            return @shape.occupy_position?(x - @x, y - @y)
        end

        def adjacent?(other)
            return false if (@x + @shape.w < other.x || other.x + other.shape.w < @x ||
                             @y + @shape.h < other.y || other.y + other.shape.h < @y)

            y0 = [@y - 1, other.y].max
            y1 = [@y + @shape.h + 1, other.y + other.shape.h].min
            dx = other.x - (@x - 1)
            return (y0...y1).any? do |y|
                (@shape.adjacent_lines[y - (@y - 1)] & (other.shape.lines[y - other.y] << dx)) != 0
            end
        end

        def overlap?(other, newx, newy)
            if @x + @shape.w <= newx || newx + other.shape.w <= @x
                return false
            end

            y0 = [@y, newy].max
            y1 = [@y + @shape.h, newy + other.shape.h].min
            (y0...y1).any? do |y|
                ((@shape.lines[y - @y] << @x) & (other.shape.lines[y - newy] << newx)) != 0
            end
        end

        def merge(other)
            # マージのみ、otherがリンクを形成していてもなにもしない
            dx = other.x - @x
            dy = other.y - @y
            @x = [@x, other.x].min
            @y = [@y, other.y].min
            @shape = @shape.concatenated(other.shape, dx, dy)
            @locked ||= other.locked
        end

        def link(other)
            @link_prev = @link_next = self if @link_next.nil?
            other.link_prev = other.link_next = other if other.link_next.nil?

            next1 = @link_next
            prev2 = other.link_prev

            @link_next = other
            next1.link_prev = prev2
            other.link_prev = self
            prev2.link_next = next1
        end

        def contains_link?(other)
            return false if @link_next.nil?
            q = self
            while (q = q.link_next) != self
                return true if q == other
            end
            return false
        end

        def add_hidden(hidden)
            @hiddens ||= []
            hidden.freeze()
            @hiddens << hidden
        end

        def remove_hidden(x, y)
            i = @hiddens.find_index {|hidden| hidden[:x] == x && hidden[:y] == y}
            unless i.nil?
                @hiddens.delete_at(i)
                @hiddens = nil if @hiddens.empty?
            end
        end

        def freeze()
            @hiddens.freeze() unless @hiddens.nil?
            super()
        end

        # frozen解除：複製したJellyを返す
        # リンクがある場合、それらも複製する
        def unfrozen()
            return self unless frozen?

            if @link_next.nil?
                cloned = self.dup()
                cloned.hiddens = @hiddens.dup() unless @hiddens.nil?
                return cloned
            end

            jelly = self
            result = nil
            while true
                cloned = jelly.dup()
                cloned.hiddens = jelly.hiddens.dup() unless jelly.hiddens.nil?
                cloned.link_next = cloned.link_prev = nil
                result.link(cloned) unless result.nil?
                result = cloned

                jelly = jelly.link_next
                break if jelly == self
            end
            return result.link_next
        end

        def inspect
            return "#<Jelly: x=#{@x}, y=#{@y}, color=#{@black_char || @color}, shape=#{@shape.inspect}#{@locked ? ' locked' : ''}#{@link_next ? ' link' : ''}#{@hiddens ? " hiddens=#{@hiddens.length}" : ''}}>"
        end

        def hash()
            return [@x, @y, @color, @shape.object_id, @hiddens].hash  # shapeが同じ形のものは同じオブジェクトなので、object_idで代用
        end

        def eql?(other)
            return @x == other.x && @y == other.y && @color == other.color &&
                   @shape == other.shape && @hiddens.hash == other.hiddens.hash
        end

        def <=>(other)
            return @y <=> other.y if @y != other.y
            return @x <=> other.x if @x != other.x
            return @shape.h <=> other.shape.h if @shape.h != other.shape.h
            @shape.w <=> other.shape.w if @shape.w != other.shape.w
        end
    end
end

File: java/jelly/Jelly.java
package jelly;

import java.util.*;

public class Jelly implements Comparable<Jelly> {
    public int x;
    public int y;
    public char color;
    public JellyShape shape;
    public boolean locked;
    public Character blackChar; // 黒ブロック時の元の文字
    public Jelly linkPrev;
    public Jelly linkNext;
    public List<Hidden> hiddens;
    private boolean frozen;

    public Jelly(int x, int y, char color, JellyShape shape, boolean locked) {
        this.x = x;
        this.y = y;
        this.color = color;
        this.shape = shape;
        this.locked = locked;
        this.linkPrev = null;
        this.linkNext = null;
        this.hiddens = null;
        this.frozen = false;
    }

    public boolean isFrozen() {
        return frozen;
    }

    public void freeze() {
        this.frozen = true;
    }

    public boolean occupyPosition(int px, int py) {
        return shape.occupyPosition(px - x, py - y);
    }

    private static int shiftLeft(int val, int shift) {
        if (shift >= 0) {
            return val << shift;
        } else {
            return val >>> -shift;
        }
    }

    public boolean adjacent(Jelly other) {
        if (this.x + this.shape.getW() < other.x || other.x + other.shape.getW() < this.x ||
            this.y + this.shape.getH() < other.y || other.y + other.shape.getH() < this.y) {
            return false;
        }

        int y0 = Math.max(this.y - 1, other.y);
        int y1 = Math.min(this.y + this.shape.getH() + 1, other.y + other.shape.getH());
        int dx = other.x - (this.x - 1);

        for (int y = y0; y < y1; y++) {
            int thisAdj = this.shape.getAdjacentLines()[y - (this.y - 1)];
            int otherLine = other.shape.getLines()[y - other.y];
            if ((thisAdj & shiftLeft(otherLine, dx)) != 0) {
                return true;
            }
        }
        return false;
    }

    public boolean overlap(Jelly other, int newx, int newy) {
        if (this.x + this.shape.getW() <= newx || newx + other.shape.getW() <= this.x) {
            return false;
        }

        int y0 = Math.max(this.y, newy);
        int y1 = Math.min(this.y + this.shape.getH(), newy + other.shape.getH());

        for (int y = y0; y < y1; y++) {
            int thisLine = this.shape.getLines()[y - this.y];
            int otherLine = other.shape.getLines()[y - newy];
            if ((shiftLeft(thisLine, this.x) & shiftLeft(otherLine, newx)) != 0) {
                return true;
            }
        }
        return false;
    }

    public void merge(Jelly other) {
        assert !frozen : "Cannot merge into a frozen Jelly";
        int dx = other.x - this.x;
        int dy = other.y - this.y;
        this.x = Math.min(this.x, other.x);
        this.y = Math.min(this.y, other.y);
        this.shape = this.shape.concatenated(other.shape, dx, dy);
        this.locked = this.locked || other.locked;
    }

    public void link(Jelly other) {
        assert !frozen : "Cannot modify links of a frozen Jelly";
        if (this.linkNext == null) {
            this.linkPrev = this;
            this.linkNext = this;
        }
        if (other.linkNext == null) {
            other.linkPrev = other;
            other.linkNext = other;
        }

        Jelly next1 = this.linkNext;
        Jelly prev2 = other.linkPrev;

        this.linkNext = other;
        next1.linkPrev = prev2;
        other.linkPrev = this;
        prev2.linkNext = next1;
    }

    public boolean containsLink(Jelly other) {
        if (this.linkNext == null) {
            return false;
        }
        Jelly q = this;
        while ((q = q.linkNext) != this) {
            if (q == other) {
                return true;
            }
        }
        return false;
    }

    public void addHidden(Hidden hidden) {
        assert !frozen : "Cannot add hidden to a frozen Jelly";
        if (this.hiddens == null) {
            this.hiddens = new ArrayList<>();
        }
        this.hiddens.add(hidden);
    }

    public void removeHidden(int hx, int hy) {
        assert !frozen : "Cannot remove hidden from a frozen Jelly";
        if (this.hiddens == null) return;
        int index = -1;
        for (int i = 0; i < this.hiddens.size(); i++) {
            Hidden h = this.hiddens.get(i);
            if (h.x == hx && h.y == hy) {
                index = i;
                break;
            }
        }
        if (index != -1) {
            this.hiddens.remove(index);
            if (this.hiddens.isEmpty()) {
                this.hiddens = null;
            }
        }
    }

    public Jelly shallowCopy() {
        Jelly cloned = new Jelly(this.x, this.y, this.color, this.shape, this.locked);
        cloned.blackChar = this.blackChar;
        return cloned;
    }

    public Jelly unfrozen() {
        if (!this.frozen) {
            return this;
        }

        if (this.linkNext == null) {
            Jelly cloned = this.shallowCopy();
            cloned.frozen = false;
            if (this.hiddens != null) {
                cloned.hiddens = new ArrayList<>();
                for (Hidden h : this.hiddens) {
                    Hidden hCopy = h.copy();
                    hCopy.jelly = cloned;
                    cloned.hiddens.add(hCopy);
                }
            }
            return cloned;
        }

        Map<Jelly, Jelly> cloneMap = new HashMap<>();
        Jelly curr = this;
        do {
            Jelly cloned = curr.shallowCopy();
            cloned.frozen = false;
            if (curr.hiddens != null) {
                cloned.hiddens = new ArrayList<>();
                for (Hidden h : curr.hiddens) {
                    Hidden hCopy = h.copy();
                    hCopy.jelly = cloned;
                    cloned.hiddens.add(hCopy);
                }
            }
            cloneMap.put(curr, cloned);
            curr = curr.linkNext;
        } while (curr != this);

        curr = this;
        do {
            Jelly cloned = cloneMap.get(curr);
            cloned.linkNext = cloneMap.get(curr.linkNext);
            cloned.linkPrev = cloneMap.get(curr.linkPrev);

            if (cloned.hiddens != null) {
                for (Hidden h : cloned.hiddens) {
                    if (h.jelly != null && cloneMap.containsKey(h.jelly)) {
                        h.jelly = cloneMap.get(h.jelly);
                    }
                }
            }
            curr = curr.linkNext;
        } while (curr != this);

        return cloneMap.get(this);
    }

    @Override
    public int compareTo(Jelly other) {
        if (this.y != other.y) {
            return Integer.compare(this.y, other.y);
        }
        if (this.x != other.x) {
            return Integer.compare(this.x, other.x);
        }
        if (this.shape.getH() != other.shape.getH()) {
            return Integer.compare(this.shape.getH(), other.shape.getH());
        }
        return Integer.compare(this.shape.getW(), other.shape.getW());
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Jelly jelly = (Jelly) o;
        if (x != jelly.x || y != jelly.y || color != jelly.color || locked != jelly.locked) {
            return false;
        }
        if (!Objects.equals(shape, jelly.shape)) {
            return false;
        }
        return Objects.equals(hiddens, jelly.hiddens);
    }

    @Override
    public int hashCode() {
        return Objects.hash(x, y, color, shape, hiddens);
    }

    @Override
    public String toString() {
        return "Jelly{x=" + x + ", y=" + y + ", color=" + (blackChar != null ? blackChar : color) + 
               ", shape=" + shape + (locked ? " locked" : "") + (linkNext != null ? " linked" : "") +
               (hiddens != null ? " hiddens=" + hiddens.size() : "") + "}";
    }
}
